add mimir support

QuantumEnigmaa opened the pull request:
### What this PR does / why we need it

Towards https://github.com/giantswarm/roadmap/issues/3217

This PR adds support for mimir in the observability-operator by allowing it to create remoteWrite secrets directing the prometheus-agents towards mimir directly and by creating the secret used by the mimir-gateway ingress for basic auth.


### Checklist

- [x] Update changelog in CHANGELOG.md.

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -124,6 +127,12 @@ func (r *ClusterMonitoringReconciler) reconcile(ctx context.Context, cluster *cl
 		return ctrl.Result{RequeueAfter: 5 * time.Minute}, errors.WithStack(err)
 	}
 
+	err = r.MimirService.ReconcileMimirConfig(ctx, r.ManagementCluster.Name)
```
Maybe inject the managementCluster name in the service?

QuantumEnigmaa commented:
Only remaining error in go code : 
```
pkg/monitoring/prometheusagent/service.go:106:18: undefined: IsMimirEnabled
```

But the thing is that this file already imports the `"github.com/giantswarm/observability-operator/pkg/monitoring"` package which includes the mimir one so it should recognize the `IsMimirEnabled` function. 
And adding another import such as this one : `"github.com/giantswarm/observability-operator/pkg/monitoring/mimir"` leads to another error : `import cycle not allowed`

QuentinBisson commented:
I Can look in 2 hours :)

QuentinBisson commented on pkg/monitoring/mimir/ingress/ingress.go:
```diff
@@ -0,0 +1,36 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/observability-operator/pkg/monitoring/mimir"
```
It's not used right?

QuantumEnigmaa replied:
That's what vscode says but the file still needs it as it implements the function as a method from the `MimirService`

QuentinBisson replied:
Oh that's because you're not in the same package when you're defining a method but you're not really using any thing from the mimirservice so you can remove this `(ms *MimirService)` and make it a function instead

QuantumEnigmaa replied:
Done

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,117 @@
+
+func (ms *MimirService) CreateOrUpdateIngressSecret(ctx context.Context, mc string, logger logr.Logger) error {
+	objectKey := client.ObjectKey{
+		Name:      "mimir-gateway-ingress",
```
I would probably extract those constants into well a const :D

QuantumEnigmaa replied:
Sure :)

QuantumEnigmaa replied:
Done

QuentinBisson commented on pkg/monitoring/mimir/utils.go:
```diff
@@ -0,0 +1,62 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	"github.com/giantswarm/observability-operator/pkg/monitoring/prometheusagent"
```
```suggestion
```

QuentinBisson replied:
So first, you're nor using it as prometheusagent.readRemoteWritePasswordFromSecret so it's not using the one you think it is and I guess your cyclic issue caused the compiler to crash, also this function is private right now (starting with a lower case letter)

QuantumEnigmaa replied:
I changed the function into a public one

QuantumEnigmaa replied:
I still have the `BuildIngressSecret` function as "undefined" though :/

QuantumEnigmaa replied:
Solved it by calling the "ingress" package

QuentinBisson commented on pkg/monitoring/prometheusagent/secret.go:
```diff
@@ -19,9 +19,18 @@ func getPrometheusAgentRemoteWriteSecretName(cluster *clusterv1.Cluster) string
+	var url string
+	var remoteWriteName string
+
+	if mimirEnabled {
```
if you're in this code, mimir is enabled, you don't need that extra check

QuantumEnigmaa replied:
Removed this as well

QuentinBisson commented on pkg/monitoring/prometheusagent/types.go:
```diff
@@ -13,8 +13,7 @@ const (
 
-	remoteWriteEndpointTemplateURL = "https://prometheus.%s/%s/api/v1/write"
-	remoteWriteName                = "prometheus-meta-operator"
+	remoteWriteEndpointTemplateURL = "https://%s.%s/%s/api/v1/%s"
```
I think most of the %s should be hardcoded as mimir will always be in the same api

QuantumEnigmaa replied:
Updated it

QuantumEnigmaa commented on pkg/monitoring/prometheusagent/service.go:
```diff
@@ -103,6 +103,7 @@ func (pas PrometheusAgentService) createOrUpdateSecret(ctx context.Context,
 		Name:      getPrometheusAgentRemoteWriteSecretName(cluster),
 		Namespace: cluster.GetNamespace(),
 	}
+	mimirEnabled := IsMimirEnabled(ctx)
```
That's not what I understood. AFAIU, this part of the code can also be used for the current setup (with prometheus instances on the MC)

QuantumEnigmaa commented on pkg/monitoring/prometheusagent/secret.go:
```diff
@@ -68,13 +77,21 @@ func (pas PrometheusAgentService) buildRemoteWriteSecret(
+	var remoteWriteName string
+
+	if mimirEnabled {
+		remoteWriteName = "mimir"
```
Removed this

QuentinBisson commented on pkg/monitoring/mimir/utils.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+// Checks whether Mimir is enabled in the cluster by listing the pods in the Mimir namespace.
+func IsMimirEnabled(ctx context.Context) bool {
```
You don't need that function then :)

QuentinBisson commented on Dockerfile:
```diff
@@ -25,7 +25,9 @@ RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o ma
 # Use distroless as minimal base image to package the manager binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+FROM alpine:3.19.1
```
Can you update the comment?

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -44,6 +45,8 @@ type ClusterMonitoringReconciler struct {
 	prometheusagent.PrometheusAgentService
 	// HeartbeatRepository is the repository for managing heartbeats.
 	heartbeat.HeartbeatRepository
+	// MimirService is the repository for managing mimir config.
```
```suggestion
	// MimirService is the service for managing mimir configuration.
```

QuantumEnigmaa replied:
Done

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -124,6 +127,12 @@ func (r *ClusterMonitoringReconciler) reconcile(ctx context.Context, cluster *cl
 		return ctrl.Result{RequeueAfter: 5 * time.Minute}, errors.WithStack(err)
 	}
 
+	err = r.MimirService.ReconcileMimirConfig(ctx, r.ManagementCluster.Name)
```
```suggestion
	err = r.MimirService.ConfigureMimir(ctx, r.ManagementCluster.Name)
```
?

QuentinBisson replied:
Btw, I would prefer if you add the management cluster in the service instead of here. WDYT?

QuantumEnigmaa replied:
I'd still keep the `reconcile` in the name because that's basically what it does

QuentinBisson replied:
That's what thé controller does, not the service :)

QuantumEnigmaa replied:
You have a point ! Updated it

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -124,6 +127,12 @@ func (r *ClusterMonitoringReconciler) reconcile(ctx context.Context, cluster *cl
 
+	err = r.MimirService.ReconcileMimirConfig(ctx, r.ManagementCluster.Name)
+	if err != nil {
+		logger.Error(err, "failed to create or updatemimir config")
```
```suggestion
		logger.Error(err, "failed to configure mimir")
```

QuantumEnigmaa replied:
Done

QuentinBisson commented on pkg/monitoring/mimir/ingress/ingress.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+const (
+	secretName      = "mimir-gateway-ingress"
```
You have this const twice right? How about adding it to pkg/monitoring/mimir/utils.go instead?

QuantumEnigmaa replied:
I moved it to the utils file but I still have to declare those consts again because I can't use these constants in the ingress file since it's in a different package

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,124 @@
+		return errors.WithStack(err)
+	}
+
+	// UPDATE SECRET
```
Do we need the upgrade part?

QuantumEnigmaa replied:
I guess we don't since in any case with the usage of `htpasswd` command we can't update the secret's value as a new hash will be generated each time.

QuantumEnigmaa commented:
operator in crashloopbackoff : 
```
2024-05-21T08:45:32Z    INFO    ensuring mimir config   {"controller": "cluster", "controllerGroup": "cluster.x-k8s.io", "controllerKind": "Cluster", "Cluster": {"name":"golem","namespace":"org-giantswarm"}, "namespace": "org-giantswarm", "name": "golem", "reconcileID": "a787843c-4
c3b-4fe4-96a1-4e0952cb96cd", "cluster": "golem"}                                                                                                                                                                                                                                          
2024-05-21T08:45:32Z    INFO    Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference        {"controller": "cluster", "controllerGroup": "cluster.x-k8s.io", "controllerKind": "Cluster", "Cluster": {"name":"golem","namespace":"org-giantswa
rm"}, "namespace": "org-giantswarm", "name": "golem", "reconcileID": "a787843c-4c3b-4fe4-96a1-4e0952cb96cd"}                                                                                                                                                                              
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                                                                                                                                       
        panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                                                           
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x15e69df]                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                          
goroutine 110 [running]:                                                                                                                                                                                                                                                                  
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()                                                                                                                                                                                                    
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:116 +0x1e5                                                                                                                                                                               
panic({0x175b9a0?, 0x28828c0?})                                                                                                                                                                                                                                                           
        /usr/local/go/src/runtime/panic.go:770 +0x132                                                                                                                                                                                                                                     
github.com/giantswarm/observability-operator/pkg/monitoring/mimir.(*MimirService).CreateOrUpdateIngressSecret(0xc00050d8d0, {0x1bfef68, 0xc00009ec90}, {0x7ffff4986756, 0x5}, {{0x1c03310?, 0xc003cb9410?}, 0x0?})                                                                        
        /workspace/pkg/monitoring/mimir/service.go:49 +0x7f                                                                                                                                                                                                                               
github.com/giantswarm/observability-operator/pkg/monitoring/mimir.(*MimirService).ReconcileMimirConfig(0xc00050d8d0, {0x1bfef68, 0xc00009ec90}, {0x7ffff4986756, 0x5})                                                                                                                    
        /workspace/pkg/monitoring/mimir/service.go:31 +0x1a5                                                                                                                                                                                                                              
github.com/giantswarm/observability-operator/internal/controller.(*ClusterMonitoringReconciler).reconcile(0xc00050d7c0, {0x1bfef68, 0xc00009ec90}, 0xc0000bed00)                                                                                                                          
        /workspace/internal/controller/cluster_monitoring_controller.go:130 +0x39c                                                                                                                                                                                                        
github.com/giantswarm/observability-operator/internal/controller.(*ClusterMonitoringReconciler).Reconcile(0xc00050d7c0, {0x1bfef68, 0xc00009ec90}, {{{0xc00014a120?, 0x5?}, {0xc00014a110?, 0xc000185d10?}}})                                                                             
        /workspace/internal/controller/cluster_monitoring_controller.go:94 +0x2ab                                                                                                                                                                                                         
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1c03310?, {0x1bfef68?, 0xc00009ec90?}, {{{0xc00014a120?, 0xb?}, {0xc00014a110?, 0x0?}}})                                                                                                                 
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:119 +0xb7                                                                                                                                                                                
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00014e460, {0x1bfefa0, 0xc0004b5cc0}, {0x17dda80, 0xc000268080})                                                                                                                                 
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:316 +0x3bc                                                                                                                                                                               
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00014e460, {0x1bfefa0, 0xc0004b5cc0})                                                                                                                                                         
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266 +0x1be                                                                                                                                                                               
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()                                                                                                                                                                                                      
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227 +0x79                                                                                                                                                                                
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 44                                                                                                                                                                               
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:223 +0x50c
```

QuantumEnigmaa commented:
After further investigations, it appears that this error is thrown by the `QueryTSDBHeadSeries` function in the pkg/monitoring/mimir/querier.go file.
After looking at grafana dashboards I noticed that, as mentioned by the error message from the operator, there were indeed no time series : all dashboards are empty. 

So I'll first need to solve this issue, meaning having timeseries ingested and accessible from mimir before moving on.

QuantumEnigmaa commented:
**Update** : 

* The operator is now able to generate the remotewrite-secret with the correct credentials for new clusters or if the secret doesn't exist for the MC 
* The only remaining issue is that the operator is unable to delete the remotewrite-secrets, whether the ones it created itself or the already existing ones to replace them

QuantumEnigmaa commented:
Once the secret deletion will be fixed, this PR will be ready for review as everything is now working :)

QuentinBisson commented on pkg/common/secret/manager.go:
```diff
@@ -0,0 +1,34 @@
+type SimpleManager struct {
+}
+
+func (m SimpleManager) GenerateGenericSecret(secretName string, secretNamespace string,
```
Do we need a manager for that? A simple function should have been enought right?

QuantumEnigmaa replied:
True, but I created a manager because I thought that this might prove useful in the future if we need to add other methods there

QuentinBisson replied:
Can we create a manager later if we need to? This feels like premature optimization that is not really helping here imo

QuantumEnigmaa replied:
Actually I found an additional usage for this manager : a `DeleteSecret` function which is really useful for deleting both mimir secrets as well as remotewrite secrets.

QuentinBisson replied:
Why is a util function not enough ? Does it requires you to create an object and pass it around?

QuantumEnigmaa replied:
Hmmm, maybe using this file as a list of utils functions might be enough indeed.

QuantumEnigmaa replied:
Done

QuentinBisson replied:
I guess it's not pushed yet :D

QuentinBisson commented on Dockerfile:
```diff
@@ -23,9 +23,10 @@ COPY internal/controller/ internal/controller/
-# Use distroless as minimal base image to package the manager binary
-# Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+# Use alpine to have the smallest image possible while being able to use a required package for the code 
```
```suggestion
# Use alpine instead of distroless to have the smallest image possible while being able to use a required package for the code (htpasswd)
# TODO(atlas): Move back to gcr.io/distroless/static:nonroot once we do not need htpasswd
```

QuentinBisson commented on main.go:
```diff
@@ -195,11 +197,18 @@ func main() {
 	}
 
+	mimirService := mimir.MimirService{
+		Client:          mgr.GetClient(),
```
```suggestion
		Client:          mgr.GetClient(),
		ManagementCluster: managementCluster,
```

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -122,8 +125,14 @@ func (r *ClusterMonitoringReconciler) reconcile(ctx context.Context, cluster *cl
 		}
 	}
 
+	err := r.MimirService.ConfigureMimir(ctx, r.ManagementCluster.Name)
```
Do we need it to run for all clusters? Maybe we could run this only for the MC reconciliation?

QuantumEnigmaa replied:
True

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+
+func (ms *MimirService) ConfigureMimir(ctx context.Context, mc string) error {
+	logger := log.FromContext(ctx).WithValues("cluster", mc)
+	logger.Info("ensuring mimir config")
```
```suggestion
	logger.Info("configuring mimir ingress")
```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+
+	err := ms.CreateAuthSecret(ctx, logger, mc)
+	if err != nil {
+		logger.Error(err, "failed to create mimit auth secret")
```
```suggestion
		logger.Error(err, "failed to create mimir auth secret")
```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+		return errors.WithStack(err)
+	}
+
+	logger.Info("ensured mimir config")
```
```suggestion
	logger.Info("configured mimir ingress")
```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+	SecretManager   secret.Manager
+}
+
+func (ms *MimirService) ConfigureMimir(ctx context.Context, mc string) error {
```
```suggestion
// ConfigureMimir configures the ingress and its authentication (basic auth) to allow prometheus agents to send their data to Mimir
func (ms *MimirService) ConfigureMimir(ctx context.Context, mc string) error {
```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+	logger := log.FromContext(ctx).WithValues("cluster", mc)
+	logger.Info("ensuring mimir config")
+
+	err := ms.CreateAuthSecret(ctx, logger, mc)
```
```suggestion
	err := ms.CreateIngressAuthenticationSecret(ctx, logger, mc)
```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+const (
+	authSecretName               = "mimir-gateway-ingress"
+	authSecretNamespace          = "mimir"
+	mimirSpecificSecretName      = "mimir-basic-auth"
```
I'm not sure about the names here.

WDYT about having:
mimirNamespace = "mimir"
mimirApiKey = "mimir-api-key"
mimirIngressAuth = "mimir-ingress-auth"

QuantumEnigmaa replied:
I agree with `mimirNamespace = "mimir"`. 
I'd still like to keep `authSecretName = "mimir-gateway-ingress"` though.
As for `mimirApiKey = "mimir-api-key"` why not :shrug:

QuentinBisson replied:
Can you name it smth Line mimir-gateway-ingress-auth then? And rename authSecretName ? I'm just thinking that auth kind means everything and nothing AT thé same time

QuantumEnigmaa replied:
Would `ingressAuthSecretName = "mimir-gateway-ingress-auth" ` work for you ?

QuentinBisson replied:
Looks good :)

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+	logger := log.FromContext(ctx).WithValues("cluster", mc)
+	logger.Info("ensuring mimir config")
+
+	err := ms.CreateAuthSecret(ctx, logger, mc)
```
How about CreateApiKey?

QuantumEnigmaa replied:
Did the change

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+		return errors.WithStack(err)
+	}
+
+	err = ms.CreateIngressSecret(ctx, mc, logger)
```
```suggestion
	err = ms.CreateIngressAuthenticationSecret(ctx, mc, logger)
```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+			return errors.WithStack(err)
+		}
+
+		secretdata := mc + ":" + password
```
Do you need to use the username here? Can secretData not be the password only?

QuantumEnigmaa replied:
You're right, no need for username. Changing this

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+	current := &corev1.Secret{}
+	err := ms.Client.Get(ctx, objectKey, current)
+	if apierrors.IsNotFound(err) {
+		// CREATE SECRET
```
```suggestion

```

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+	return nil
+}
+
+func (ms *MimirService) DeleteIngressSecret(ctx context.Context) error {
```
Did you not forget to delete the api key secret?

QuantumEnigmaa replied:
True, hence my idea of using a `DeleteSecret` function defined in the secret manager to easily handle the deletion of both secrets

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	// Delete the finalizer
+	desired := current.DeepCopy()
```
I think the patch is not needed

QuantumEnigmaa replied:
Yes, removed it

QuentinBisson commented on pkg/monitoring/prometheusagent/secret.go:
```diff
@@ -1,27 +1,65 @@
 
-	"github.com/giantswarm/observability-operator/pkg/monitoring"
+const (
+	authSecretName      = "mimir-basic-auth"
```
All the more reason to call this api key :D

QuantumEnigmaa replied:
:shrug:

QuentinBisson commented on pkg/monitoring/prometheusagent/secret.go:
```diff
@@ -1,27 +1,65 @@
+	remoteWriteName     = "mimir"
 )
 
+func GetMimirIngressPassword(ctx context.Context) (string, error) {
```
Maybe this function should be in the mimir package and this rely on mimir instead or mimir rely on this?

QuantumEnigmaa replied:
That's what I did first but this came with import issues on the prometheusagent package :/

QuentinBisson commented on pkg/monitoring/prometheusagent/secret.go:
```diff
@@ -68,21 +103,15 @@ func (pas PrometheusAgentService) buildRemoteWriteSecret(
-			return rw.Password, nil
-		}
-	}
+	password := strings.Split(secretData, ":")[1]
```
Not prefixing with the username would remove this need

QuantumEnigmaa replied:
yes, updated it alongside the removal of the username

QuentinBisson commented on pkg/monitoring/prometheusagent/types.go:
```diff
@@ -13,8 +13,7 @@ const (
 
-	remoteWriteEndpointTemplateURL = "https://prometheus.%s/%s/api/v1/write"
-	remoteWriteName                = "prometheus-meta-operator"
+	remoteWriteEndpointTemplateURL = "https://mimir.%s/api/v1/push"
```
Why did you move the remoteWriteName over?

QuantumEnigmaa replied:
That was a very early change, before knowing that the operator would be used only when mimir would be enabled. Reverting it :)

QuentinBisson commented:
@QuantumEnigmaa you're doing a really great job so far :) i'm sorry if I have too strong of an opinion sometimes :)

QuantumEnigmaa commented:
> i'm sorry if I have too strong of an opinion sometimes :) 

No problem ! That's how I can improve myself so I appreciate it :)

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -142,6 +151,12 @@ func (r *ClusterMonitoringReconciler) reconcileDelete(ctx context.Context, clust
 				return ctrl.Result{RequeueAfter: 5 * time.Minute}, errors.WithStack(err)
 			}
+
+			err = r.MimirService.DeleteMimirSecrets(ctx)
```
So now, we probably either should have CreateMimirSecrets and DeleteMimirSecrets or ConfigureMimir and UnconfigureMimir right? not a blocker though

QuantumEnigmaa replied:
You're right. Removed it

QuentinBisson commented on pkg/common/secret/utils.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+func GenerateGenericSecret(secretName string, secretNamespace string,
+	key string, value string) (*corev1.Secret, error) {
```
If you're not running the create, why do you need an error?

QuantumEnigmaa replied:
You're right. Removed it

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,144 @@
+// ConfigureMimir configures the ingress and its authentication (basic auth)
+// to allow prometheus agents to send their data to Mimir
+func (ms *MimirService) ConfigureMimir(ctx context.Context) error {
+	logger := log.FromContext(ctx).WithValues("cluster", ms.ManagementCluster.Name)
```
Could we add the mc name to the context instead? That way we do not have to pass the logger around to subfunctions

QuentinBisson replied:
Also, I'm not sure the mc name brings a lot of value so you can probably use `log.FromContext(ctx)` directly

QuentinBisson replied:
@QuantumEnigmaa Can you check this one real quick? Then we're good :)

QuantumEnigmaa replied:
> Could we add the mc name to the context instead?

I'm not sure how to do that though :/

QuentinBisson replied:
```go
bg := context.Background()
ctx := context.WithValue(bg, "k", "zabba")
```

QuentinBisson replied:
So you can do this at the top of the controller I think?
ctx = context.WithValue(ctx, "cluster", cluster.GetName())
ctx = context.WithValue(ctx, "installation", mcName)

and then we should have this in all log lines. Would be nice to test :)

QuantumEnigmaa replied:
So this had to be done in the controller just before calling the `ConfigureMimir` function then right ?

QuentinBisson replied:
I think at the top of the reconcile function even :)

QuentinBisson replied:
Did you have time to check this one out?

QuantumEnigmaa replied:
No my bad I totally forgot about this

QuantumEnigmaa replied:
Done

QuentinBisson commented on pkg/common/secret/utils.go:
```diff
@@ -0,0 +1,51 @@
+	}
+	current := &corev1.Secret{}
+	// Get the current secret if it exists.
+	err := providedClient.Get(ctx, objectKey, current)
```
I've seen upstream that you can simplify the delete by removing the Get like so:

```go
if err := client.Delete(ctx, &obj); client.IgnoreNotFound(err) != nil {
  return errors.WithStack(err)
}
return nil
```

QuentinBisson replied:
Then maybe we do not need a util function to replace this one liner :)

TheoBrigitte replied:
Agreed, there's no need to get in order to delete a resource.

QuantumEnigmaa replied:
I simplified the deletion by getting rid of the get but I couldn't go as far as the one-liner example that you gave :/ 
Also I kept the secret deletion function to avoid code duplication.

WDYT ?

TheoBrigitte replied:
That's fine with me

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -60,6 +61,30 @@ func (ms *MimirService) CreateApiKey(ctx context.Context, logger logr.Logger) er
+			return errors.WithStack(err)
+		}
+
+		workloadCLustersList := &clusterv1.ClusterList{}
```
I think you just mean clusters right?

QuantumEnigmaa replied:
Right, updating it

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -60,6 +61,30 @@ func (ms *MimirService) CreateApiKey(ctx context.Context, logger logr.Logger) er
 	current := &corev1.Secret{}
 	err := ms.Client.Get(ctx, objectKey, current)
 	if apierrors.IsNotFound(err) {
+		// First all secrets using the password from the mimirApiKey secret are deleted
+		// to ensure that they won't use an outdated password.
+		logger.Info("Deleting old secrets")
```
How do you make sure the secret is recreated before the next reconciliation?

QuantumEnigmaa replied:
Well, since this deletion is happening in the case where the basic-auth secret isn't present on the cluster and considering that the remotewrite secrets and ingress secret all need the basic-auth secret's data and can't be created without it, as long as this process is not finished and the basic-auth created, they won't be created before.

Not sure if I explained it clearly :sweat_smile:

QuentinBisson replied:
alright :)

QuentinBisson commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,163 @@
+			return errors.WithStack(err)
+		}
+
+		clustersList := &clusterv1.ClusterList{}
```
Why clustersList and not Cluster? Because it's either clusterList or clusters :D

QuantumEnigmaa replied:
My bad, typo :sweat_smile:

QuantumEnigmaa commented:
Tested and working :) 
Secret rotation is fairly easy now

TheoBrigitte commented on .golangci.yml:
```diff
@@ -16,6 +16,12 @@ issues:
       linters:
         - dupl
         - lll
+    - path: "pkg/monitoring/prometheusagent/secret.go"
+      linters:
+        - gosec
+    - path: "pkg/monitoring/mimir/service.go"
+      linters:
+        - gosec
```
What nasty things are we doing there which requires to ignore gosec ?

QuantumEnigmaa replied:
gosec was complaining about const names such as `ingressAuthSecretName`

TheoBrigitte replied:
Then please fix the warning rather than ignoring the whole file. We want to keep `gosec` checking for potential security issues.

QuentinBisson replied:
We can't really fix them all because if a variable name contains secret or api key it will complain but we Can use a single Line comment over excluding a full file

QuentinBisson replied:
Like this : https://github.com/giantswarm/observability-operator/blob/ec1abdec2ce3c7a4a95c2faeb7cd431f25dec52d/main.go#L69

QuantumEnigmaa replied:
Oh nice ! I'm gonna do that then :)

TheoBrigitte commented on Dockerfile:
```diff
@@ -23,9 +23,11 @@ COPY internal/controller/ internal/controller/
-# Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+# Use alpine instead of distroless to have the smallest image possible while being able to use a required package for the code (htpasswd)
+# TODO(atlas): Move back to gcr.io/distroless/static:nonroot once we do not need htpasswd
```
Is there a point in the future where we would not need htpasswd ?

QuantumEnigmaa replied:
Yes, most likely when multi-tenancy will be an actual thing as we'll rely on the multi-tenant proxy for authentication, if I understood correctly.

QuentinBisson replied:
I think we can Switch back hère as well if that works :)

TheoBrigitte commented on pkg/monitoring/mimir/service.go:
```diff
@@ -0,0 +1,163 @@
+
+		for _, cluster := range clusterList.Items {
+			secretName := prometheusagent.GetPrometheusAgentRemoteWriteSecretName(&cluster)
+			err = secret.DeleteSecret(secretName, cluster.Namespace, ctx, ms.Client)
```
How do we make sure those secrets are not re-created afterwards ?

QuantumEnigmaa replied:
Check what I answered Quentin ;) 

> Well, since this deletion is happening in the case where the basic-auth secret isn't present on the cluster and considering that the remotewrite secrets and ingress secret all need the basic-auth secret's data and can't be created without it, as long as this process is not finished and the basic-auth created, they won't be created before.
Not sure if I explained it clearly 😅

TheoBrigitte replied:
I did not understand that ^

QuantumEnigmaa replied:
i'll try to rephrase : 

basically, remotewrite secrets as well as the secret used by the ingress rely on the mimirApiKey secret's data (whose value is `username:password`) for their creation process meaning that as long as the mimirApiKey secret is not created, neither can they. 
Since the piece of code you mentioned is happening **prior** to the mimirApiKey secret being created, one can be sure that all other secrets won't be created before this piece of code has been executed.

TheoBrigitte replied:
If I got this right, here we are deleting the remotewrite secret, and mimirApiKey creation happens later on right?
What happen then when we delete the secret here and the mimiApiKey creation is triggered ? Isn't this causing a failure ?

TheoBrigitte commented on pkg/common/secret/utils.go:
```diff
@@ -0,0 +1,51 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func GenerateGenericSecret(secretName string, secretNamespace string,
+	key string, value string) *corev1.Secret {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: secretNamespace,
+		},
+		Data: map[string][]byte{
+			key: []byte(value),
+		},
+		Type: "Opaque",
+	}
+
+	return secret
+}
```
I would prefer that we remove this function to avoid obfuscation and directly create the secret object where needed.

QuantumEnigmaa replied:
But that would lead to a lot of duplicate code :/

TheoBrigitte replied:
I am not too strong on this one, do what you feel is right. I expressed my concern.

TheoBrigitte commented on pkg/monitoring/prometheusagent/secret.go:
```diff
@@ -63,21 +101,13 @@ func (pas PrometheusAgentService) buildRemoteWriteSecret(
+func readMimirAuthPasswordFromSecret(secret corev1.Secret) (string, error) {
+	var secretData string
+
+	err := yaml.Unmarshal(secret.Data["credentials"], &secretData)
```
We should check for existence of the `credentials` key here otherwise this would panic.

QuantumEnigmaa replied:
Agreed

TheoBrigitte reviewed (commented):
Very nice, but I have few remarks if you may Sir.

TheoBrigitte commented on pkg/common/password/manager.go:
```diff
@@ -19,3 +22,12 @@ func (m SimpleManager) GeneratePassword(length int) (string, error) {
 }
+
+func (m SimpleManager) GenerateHtpasswd(username string, password string) (string, error) {
+	htpasswd, err := exec.Command("htpasswd", "-bn", username, password).Output()
+	if err != nil {
+		return "", err
+	}
+	formattedHtpasswd := strings.TrimSpace(string(htpasswd))
+	return formattedHtpasswd, nil
```
We can probably get rid of the `htpasswd` dependency and use something like


```suggestion
	encryptedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return "", err
	}
	formattedHtpasswd := fmt.Sprintf("%s:%s", username, string(encryptedPassword))
	return formattedHtpasswd, nil
```

Where `bcrypt` is `"golang.org/x/crypto/bcrypt"`

Need to be tested

QuantumEnigmaa replied:
I'm gonna try this without the username being added as well so that it stays compliant with the rest of the code

TheoBrigitte replied:
Beware that `htpasswd` does add the username in front of the hashed password, producing a string in the following format:
```
user:password_hash
```

I kept the same format here when using `bcrypt`.

QuantumEnigmaa replied:
Oooo because `bcrypt` is only creating a salted hash for the password right ?

TheoBrigitte replied:
Yes it is

QuantumEnigmaa replied:
Ok makes sense then :+1:

QuantumEnigmaa replied:
So I tried it on `golem` and from what I see it seems to be working :)

TheoBrigitte replied:
:tada:

QuentinBisson commented on pkg/common/password/manager.go:
```diff
@@ -3,10 +3,14 @@ package password
 
 type Manager interface {
 	GeneratePassword(length int) (string, error)
+	GenerateHtpasswd(username string, password string) (string, error)
```
You can improve liké that if you like
```suggestion
	GenerateHtpasswd(username, password string) (string, error)
```

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -83,25 +81,29 @@ func (r *ClusterMonitoringReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		return ctrl.Result{}, errors.WithStack(err)
 	}
 
+	ctx = context.WithValue(ctx, "cluster", cluster.Name)
```
Can you test this to make sure this logs right?

QuantumEnigmaa replied:
Sure

QuentinBisson replied:
It's missing the installation though

QuantumEnigmaa replied:
True, there is no sign of it in any of the operator's logs

QuantumEnigmaa replied:
Should I revert back to the previous code ?

QuentinBisson replied:
I'm not sure we really need the installation so I'd leave as is?

QuentinBisson replied:
I have an idea :D

QuentinBisson commented on internal/controller/cluster_monitoring_controller.go:
```diff
@@ -80,25 +81,29 @@ func (r *ClusterMonitoringReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		return ctrl.Result{}, errors.WithStack(err)
 	}
 
+	ctx = context.WithValue(ctx, "cluster", cluster.Name)                  // nolint
```
Can you explain why the nolint?

QuantumEnigmaa replied:
I'm gonna add a comment

QuantumEnigmaa replied:
Done

QuentinBisson replied:
Can you try:

```go
logger := log.FromContext(ctx)
                       .WithValues("cluster", cluster.Name)
                       .WithValues("installation", r.ManagementCluster.Name)
ctx = log.IntoContext(ctx, logger)
```

QuentinBisson replied:
I think we have a bit too many clusters but we can sort this out later :)

QuentinBisson reviewed (approved):
LGTM. @TheoBrigitte do you want to take another look?

marieroque commented:
It would be great having the schema you made during your presentation in the README to see communication between prometheus-agent, mimir-gateway and the creation of the different secrets on the MC :-)

QuantumEnigmaa commented:
> It would be great having the schema you made during your presentation in the README to see communication between prometheus-agent, mimir-gateway and the creation of the different secrets on the MC :-)

I was planning to add it on our intranet doc :shrug:

QuantumEnigmaa commented:
@QuentinBisson @marieroque @TheoBrigitte could you please give a final review ? :pray: 
I think I addressed most of the suggestions and remarks for this PR.
